clash-lib/tests/Clash/Tests/Driver/Manifest.hs

```diff
@@ -0,0 +1,51 @@
+module Clash.Tests.Driver.Manifest where
+
+import qualified Data.Aeson as Aeson
+import qualified Data.Aeson.Encode.Pretty as Aeson
+import Data.ByteString (ByteString)
+import qualified Data.ByteString.Base16 as Base16
+import Data.Coerce (coerce)
+import Data.Text (Text)
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+
+import Test.Tasty
+
+import qualified Test.Tasty.QuickCheck as Q
+import qualified Test.QuickCheck.Utf8 as Q
+
+import Clash.Driver.Manifest
+
+newtype ArbitraryText = ArbitraryText Text deriving (Show)
+newtype ArbitraryManifest = ArbitraryManifest Manifest deriving (Show)
+
+instance Q.Arbitrary ArbitraryText where
+  arbitrary = coerce Q.genValidUtf8
+  shrink = coerce Q.shrinkValidUtf8
+
+genDigest :: Q.Gen ByteString
+genDigest = Base16.encode . Text.encodeUtf8 . coerce @ArbitraryText <$> Q.arbitrary
+
+genFilePath :: Q.Gen FilePath
+genFilePath = Text.unpack . coerce @ArbitraryText <$> Q.arbitrary
+
+genManifest :: Q.Gen Manifest
+genManifest =
+  Manifest
+    <$> Q.arbitrary -- hash
+    <*> Q.arbitrary -- flags
+    <*> fmap (map (coerce @ArbitraryText @Text)) Q.arbitrary -- in ports
```

Comment:
Do you need to `map` the `coerce` here? I thought for many types `fmap coerce == coerce` meaning

```haskell
fmap coerce (Q.arbitrary @ArbitraryText)
```

would suffice

Reply:
You're absolutely right!

Reply:
do you even need to `fmap` then? won't the following work?:
```haskell
coerce @(Q ArbitraryText) @(Q Text)
```

Reply:
assuming the role of the type parameter of whatever applicative you're in is `representational`

Reply:
Oh good point, I didn't even think about type roles. Yes you should be able to write

```haskell
coerce (Q.arbitrary @ArbitraryText)
```

Reply:
We wouldn't get anything done without type checkers ;-)

```haskell
coerce @(Q.Gen [ArbitraryText]) @(Q.Gen [Text])
```